### DISCUSSION TITLE
Numbering scheme for "due" for new cards

### DISCUSSION THREAD

**Discussion by xofm31** [mortii/anki-morphs#167] [Ideas]
I would like to put forward for your consideration two suggestions for the due number for new cards.

1. Change the number for a new card that has no morphs to something lower than the highest possible number: change `    default_difficulty = 2147483647` to something like ` default_difficulty = 9999999`. When a card gets repositioned, and all of the cards get moved up, they wind up with a negative number, which puts them at the top of the queue. This happens whenever I reposition on Ankidroid, and today it just happened when I hit the Ankimorph "Learn card now" button.

2. Not reschedule all of the cards so that they have a unique due, but instead allow them to have a due number that is defined by the difficulty. This is inspired by Morphman's numbering system, where all of the cards with one unknown morph are in the 100,000's, all of the cards with two unknown morphs are in the 200,000's, and so on. To me, this has a couple of benefits:
- When I'm looking for all of the cards with the same morph, I don't need to make a distinction between the ready cards with that morph and all cards with the same morph, because it is super easy to see which is which by the due time.
- I think this should mean that if you recalc with the same frequency list, most of the cards should keep the same difficulty score. This would result in a lot of cards not having to be uploaded each time you sync. As someone who uses Ankidroid a lot, I am constantly syncing, and now that my collection is getting larger, the time required for the sync is starting to get longer.
I haven't tried it out, but I believe that if you make your `morph_unknown_penalty` to be `100000` or another multiple of 10, that might be all you would need to get that result.

It could be that these are both bad ideas for some technical reason, but I thought I'd put them out there.

**Comment by mortii** [comment]
> 1. Change the number for a new card that has no morphs to something lower than the highest possible number: change     default_difficulty = 2147483647 to something like  default_difficulty = 9999999. When a card gets repositioned, and all of the cards get moved up, they wind up with a negative number, which puts them at the top of the queue. This happens whenever I reposition on Ankidroid, and today it just happened when I hit the Ankimorph "Learn card now" button.

Yikes, I haven't thought about that. You are right, It it should definitely be lower. I'll make a new issue for it.

> 2. Not reschedule all of the cards so that they have a unique due [...]

Cards having a unique due is definitely a necessity, for better or worse. I assume you read the code comment, but for completeness sake it says this:

```
# When multiple cards have the same due (difficulty), then anki
# chooses one for review and ignores the others, [...]
```

A lot of cards can end up having the same difficulty, especially early on. This problem might not be very pronounced if you use all the 'skip' options, but some people don't use them at all, and then cards just seemingly disappear, which is super bad. 

> As someone who uses Ankidroid a lot, I am constantly syncing, and now that my collection is getting larger, the time required for the sync is starting to get longer.

I empathize, long sync times _**really**_ suck. Unfortunately, I can't think of an clean way of reducing it while still maintaining unique dues...

**Reply by xofm31** [reply]
Thank you!

> Cards having a unique due is definitely a necessity, for better or worse. I assume you read the code comment, but for completeness sake it says this:

Yes, I did see that comment, but haven't experienced it personally. I'm pretty sure I've had multiple cards with the same difficulty in non-Morphman collections, and I didn't notice the cards disappearing. I do use all the 'skip' options, but this should still be a problem, because Morphman does create cards with different morphs with the same due number. I'll have to try it out.

If the sync times get to be a real problem (they aren't for me), I suppose you could do your difficulty calculations, and then only move the first X (100? 1000?) cards to the end of the queue, leaving the rest in the situation with multiple cards with the same due. Someone would have to recalc to bring down the next set of new cards.

**Reply by mortii** [reply]
The unique due problem is very sneaky, I was very confused when I encountered it in morphman.

> I suppose you could do your difficulty calculations, and then only move the first X (100? 1000?) cards to the end of the queue, leaving the rest in the situation with multiple cards with the same due

Do you mean only giving the first X cards a unique due? That could maybe work... it might be overkill always give a unique due to any cards after the first 5K or something.

**Reply by xofm31** [reply]
> The unique due problem is very sneaky, I was very confused when I encountered it in morphman.

I just tried with about 10 cards with the same due, and it went through all of them for me. But maybe my setup is somehow different.

> Do you mean only giving the first X cards a unique due? That could maybe work... it might be overkill always give a unique due to any cards after the first 5K or something.

Yes, that is what I was suggesting could be a workaround, if you wanted a way to reduce the number of cards that had to be sync'd because they had been changed. 

Unrelated, but not worth submitting a new issue. The links on https://mortii.github.io/anki-morphs/user_guide/intro.html are broken.

**Reply by mortii** [reply]
> I just tried with about 10 cards with the same due, and it went through all of them for me. But maybe my setup is somehow different.

Wow, I can't reproduce it anymore either... it must have been fixed sometime within the last year without me noticing it. This changes everything....

> Unrelated, but not worth submitting a new issue. The links on https://mortii.github.io/anki-morphs/user_guide/intro.html are broken.

Mdbook has a fun race condition where there is a high probability that those links are broken when it is rebuilt, which happens on every push to main. I'll try to fix this once and for all.

**Comment by mortii** [comment]
Your proposed changes have been implemented as of [v1.2.8](https://github.com/mortii/anki-morphs/releases/tag/v1.2.8), so I'll close this discussion now. 

Thanks!

**Comment by xofm31** [comment]
> This happens whenever I reposition on Ankidroid, and today it just happened when I hit the Ankimorph "Learn card now" button.

Now that I'm thinking about it more, I don't think you'd want it to reposition all of your other cards when you hit "Learn card now", and when I manually reposition, I have the choice. I think this is the relevant call:
`mw.col.sched.reposition_new_cards(selected_cards, 0, 1, False, True)`
 I'm not going to track down what that is calling, but based on the manual reposition dialog box, I bet that it won't reposition all the other cards if you change the last "True" to "False". I don't know what the behavior will be if you have multiple cards selected, though - will they all be "Learn Now" or only the first if you don't move other cards out of the way.

**Reply by mortii** [reply]
Great catch, that was made with unique dues in mind. 

> `mw.col.sched.reposition_new_cards(selected_cards, 0, 1, False, True)`

Not using keyword arguments here is horrible, sorry about that, pycharm spoils me. The last argument is `shift_existing`, so you are right about that too.

> I don't know what the behavior will be if you have multiple cards selected, though - will they all be "Learn Now" or only the first if you don't move other cards out of the way.

I just tested really quickly by shifting two groups of cards to the same starting position, and as one would expect, they interweave with the existing cards. 

![Screenshot from 2024-03-04 01-28-59](https://github.com/mortii/anki-morphs/assets/15674619/ce578115-96c4-4265-bd8f-06739435715f)


This effectively makes it not a _true_ 'learn now' feature since it doesn't guarantee an earlier due than other new cards. I don't have a strong opinion about this right now, but I'm leaning towards that being okay--making sure it is a super strict feature probably doesn't outweigh the benefits of an easy heuristic that doesn't cause long sync times.

What do you think?

**Reply by xofm31** [reply]
I think either way would be fine. Another option would be to check if there is just 1 selected card, then shift_existing is False, and if it's more than 1, shift_existing is True.

**Reply by mortii** [reply]
Hmm, maybe. On the other hand it could be confusing to users that the behavior is different depending on how many cards they select. I'm not a huge fan of inconsistent effects like that.

Btw, I didn't notice a sizable reduction in the amount of synced cards today after I recalc'ed, so I might have botched the implementation in the latest update. Was it any better for you?

**Reply by xofm31** [reply]
> Hmm, maybe. On the other hand it could be confusing to users that the behavior is different depending on how many cards they select. I'm not a huge fan of inconsistent effects like that.

Fair. I think then it makes sense to ensure that "right now" means "right now". 

> Btw, I didn't notice a sizable reduction in the amount of synced cards today after I recalc'ed, so I might have botched the implementation in the latest update. Was it any better for you?

Yes, when I tested it out, it seemed to make a big difference. But I'm still using Morphman as my primary, so I haven't used it much. I'm pretty busy right now, but my plans over the next couple of weeks are (1) compare the morphs I get from Jieba (which is what Morphman uses) to spaCy for Chinese, and then (2) work on a proposal for a "card usefulness" measure that focuses on learning useful words. I'll check the number of cards syncing again when I get a chance.

**Reply by mortii** [reply]
The reason I didn't notice a difference after the 'due' change was coincidentally that I also used 'learn now', which leads to a full collection sync. When I don't use learn now, then only a couple hundred cards are uploaded instead of 50K+.

I had forgotten how nice it was to have so few cards synced, so now I actually feel more strongly that we _should_ make the learn now feature less strict to avoid massive syncs. New cards are (always?) appended, so position zero shouldn't really be occupied, making the overlap problem hopefully negligible .

**Reply by xofm31** [reply]
>  The reason I didn't notice a difference after the 'due' change was coincidentally that I also used 'learn now', 

That'll do it.

> so now I actually feel more strongly that we should make the learn now feature less strict to avoid massive syncs

Another option I thought of was maybe you could loop through all of the selected cards and reposition each of them to due 0. As long as you're not repositioning the cards, that ought to put them all ahead of any existing cards, with the possible exception of one that was already in due 0. I haven't tried to figure out what order anki gives cards that all have the same due date. Based on my Morphman recalc'ed deck, it looks like it might be by creation date.

**Reply by mortii** [reply]
> Another option I thought of was maybe you could loop through all of the selected cards and reposition each of them to due 0. 

That's a great idea, we should definitely do that. 

> I haven't tried to figure out what order anki gives cards that all have the same due date. Based on my Morphman recalc'ed deck, it looks like it might be by creation date.

That sounds about right. That being said, if someone uses 'learn now' on multiple cards at the same time, then they probably won't be too concerned with the sub-order of those cards. If they really want them in a specific order then they should use it on one card at a time.

I'll probably implement this sometime today and I'll add you as a co-author.
